Adding CVT Model

AnugunjNaman opened the pull request:
# What does this PR do?
Add CvT Model for Vision Classification

Fixes #13158


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@NielsRogge @LysandreJik

AnugunjNaman committed:
Adding cvt files

AnugunjNaman committed:
Adding cvt files

AnugunjNaman committed:
Adding cvt files

AnugunjNaman committed:
Adding cvt files

AnugunjNaman committed:
Adding cvt files

AnugunjNaman committed:
Adding cvt files

ayush1399 committed:
changes in init file

ayush1399 committed:
changes in init file

ayush1399 committed:
changes in init file

AnugunjNaman committed:
Merge branch 'cvtmodel' of https://github.com/AnugunjNaman/transformers into cvtmodel

AnugunjNaman committed:
Style fixes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17253). All of your documentation changes will be reflected on that endpoint.

AnugunjNaman commented:
@NielsRogge can you review it and suggest any further changes.
I'm not sure how you want to modify `cls_token` part. Since modifying in `CvtStage` section (stopping the split) will change shape of hidden states (stored in all hidden states and passing of 4D shape for CNN in next layer).

I leave that part to you on how you want to change it to pass it over different classes further.

I have run the `make fix-copies` and done docstrings part. I think everything is done apart from change you wanted to make for `cls_token`.

cloudhan committed:
Fix obvious typos in flax decoder impl (#17279)

Change config.encoder_ffn_dim -> config.decoder_ffn_dim for decoder.

gante committed:
TF - Fix convnext classification example (#17261)

stas00 committed:
[WIP] [doc] performance/scalability revamp (#15723)

* [doc] performance/scalability revamp

* link the new docs

* no :

* mixed precision

* work on the first doc

* expand the main doc

* Trigger CI

* style

* revamp single GPU training section

* work on training performance

* remove files not used anymore or will be added later

* final touches

* fix rebase

* Add hardware section to toctree

* fix toctree again

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* remove `fast_tokenizers` entry that was copied in rebase

* add warning about DP vs DDP

* remove todo

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* fix missing closure of codeblock

* Update docs/source/en/perf_train_gpu_many.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* sync with #16860

* update toc

Co-authored-by: leandro <leandro.vonwerra@spoud.io>
Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

KennethEnevoldsen committed:
fixed bug in run_mlm_flax_stream.py (#17203)

* fixed bug run_mlm_flax_stream.py

Fixed bug caused by an update to tokenizer keys introduced in recent transformers versions (between `4.6.2` and `4.18.0`) where additional keys were introduced to the tokenizer output.

* Update run_mlm_flax_stream.py

* adding missing paranthesis

* formatted to black

* remove cols from dataset instead

* reformat to black

* moved rem. columns to map

* formatted to black

Co-authored-by: KennethEnevoldsen <kennethcenevolsen@gmail.com>

cavdard committed:
Updated checkpoint support for Sagemaker Model Parallel (#17219)

* adding partial checkpoint support for optimizer state

* formatted trainer.py

* Refactoring based on comments

* reformatting

* Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: Cavdar <dcavdar@a07817b12d7e.ant.amazon.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

loubnabnl committed:
Update codeparrot data preprocessing (#16944)

* add new preprocessing arguments

* add new filters

* add new filters to readme

* fix config and test count, update function names and docstrings

* reformat code

* update readme

* Update readme

* rename config_test filter

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* rename few_assignments filter

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* rename tokenizer in arguments

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* rename functions and add limit_line argument for config_test filter

* update threshold for config_test filter

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>
Co-authored-by: Loubna ben allal <loubnabenallal@gmail.com>

loubnabnl committed:
CodeParrot data pretokenization (#16932)

* add pretokenization arguments

* add pretokenization script

* add support for pretokenized data

* reformat code

* fix run command for training

* fix model call from config

* remove a package

* add comments on pretokenization in the readme

* remove explicit parallelization

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* update readme

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* update readme -remove username

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* update readme -remove username

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* keep data parallelization

* reformat code

* reformat code

* update readme

* reformat code

* Update examples/research_projects/codeparrot/README.md

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>
Co-authored-by: Loubna ben allal <loubnabenallal@gmail.com>

lewtun committed:
Remove next sentence prediction from supported ONNX tasks (#17276)

MichelBartels committed:
Align logits and labels in OPT (#17237)

orieg committed:
Mlflowcallback fix nonetype error (#17171)

* Fix edge cases TypeError: 'NoneType' object is not callable

* fix style

sgugger committed:
Automatically sort auto mappings (#17250)

* Automatically sort auto mappings

* Better class extraction

* Some auto class magic

* Adapt test and underlying behavior

* Remove re-used config

* Quality

ydshieh committed:
Make TrainerHyperParameterSigOptIntegrationTest slow test (#17288)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Better error in the Auto API when a dep is missing (#17289)

ydshieh committed:
Fix FlavaForPreTrainingIntegrationTest CI test (#17232)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Use the PR URL in CI report (#17269)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sanderland committed:
logging documentation update (#17174)

* logging documentation

* style

Co-authored-by: Sander Land <sander@chatdesk.com>

k-zehnder committed:
docs(transformers): fix typo (#17263)

amyeroberts committed:
Add Tensorflow Swin model (#16988)

Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

patrickvonplaten committed:
[Tests] Fix slow opt tests (#17282)

* fix opt tests

* remove unused tok

* make style

* make flake8 happy

* Update tests/models/opt/test_modeling_opt.py

lkm2835 committed:
Fix test_model_parallelization (#17249)

* Fix test_model_parallelization

* Modify

patrickvonplaten committed:
Add Wav2Vec2Conformer (#16812)

* save intermediate

* add wav2vec2 conformer

* add more code

* more

* first test passes

* make all checkpoints work

* update

* up

* more clean ups

* save clean-up

* save clean-up

* save more

* remove bogus

* finalize design conformer

* remove vision

* finish all tests

* more changes

* finish code

* add doc tests

* add slow tests

* fix autoconfig test

* up

* correct docstring

* up

* update

* fix

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Anton Lozhkov <aglozhkov@gmail.com>

* Update docs/source/en/model_doc/wav2vec2-conformer.mdx

* upload

* save copied from

* correct configs

* fix model outputs

* add to docs

* fix imports

* finish

* finish code

* correct copied from

* correct again

* correct make fix

* improve make fix copies

* save

* correct fix copy from

* correct init structure

* correct

* fix import

* apply suggestions

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Anton Lozhkov <aglozhkov@gmail.com>

ydshieh committed:
Fix missing job action button in CI report  (#17270)

* use matrix.machine_type

* fix job names used in job_link

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Fix wrong PT/TF categories in CI report (#17272)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

NielsRogge commented on README.md:
```diff
@@ -249,6 +249,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[ConvBERT](https://huggingface.co/docs/transformers/model_doc/convbert)** (from YituTech) released with the paper [ConvBERT: Improving BERT with Span-based Dynamic Convolution](https://arxiv.org/abs/2008.02496) by Zihang Jiang, Weihao Yu, Daquan Zhou, Yunpeng Chen, Jiashi Feng, Shuicheng Yan.
 1. **[CPM](https://huggingface.co/docs/transformers/model_doc/cpm)** (from Tsinghua University) released with the paper [CPM: A Large-scale Generative Chinese Pre-trained Language Model](https://arxiv.org/abs/2012.00413) by Zhengyan Zhang, Xu Han, Hao Zhou, Pei Ke, Yuxian Gu, Deming Ye, Yujia Qin, Yusheng Su, Haozhe Ji, Jian Guan, Fanchao Qi, Xiaozhi Wang, Yanan Zheng, Guoyang Zeng, Huanqi Cao, S ...
 1. **[CTRL](https://huggingface.co/docs/transformers/model_doc/ctrl)** (from Salesforce) released with the paper [CTRL: A Conditional Transformer Language Model for Controllable Generation](https://arxiv.org/abs/1909.05858) by Nitish Shirish Keskar*, Bryan McCann*, Lav R. Varshney, Caiming Xiong and Richard Socher.
+1. **[CvT](https://huggingface.co/docs/transformers/main/model_doc/cvt)** (from Microsoft) released with the paper [CvT: Introducing Convolutions to Vision Transformers](https://arxiv.org/abs/1909.05858) by Nitish Shirish Keskar*, Bryan McCann*, Lav R. Varshney, Caiming Xiong and Richard Socher.
```
```suggestion
1. **[CvT](https://huggingface.co/docs/transformers/main/model_doc/cvt)** (from Microsoft) released with the paper [CvT: Introducing Convolutions to Vision Transformers](https://arxiv.org/abs/1909.05858) by Nitish Shirish Keskar, Bryan McCann, Lav R. Varshney, Caiming Xiong and Richard Socher.
```

NielsRogge commented on docs/source/en/index.mdx:
```diff
@@ -50,7 +50,6 @@ The library currently contains JAX, PyTorch and TensorFlow implementations, pret
 ### Supported models
 
 <!--This list is updated automatically from the README with _make fix-copies_. Do not update manually! -->
-
```
Not sure what happened here, but normally there should only be 2 added lines in index.mdx (namely, the new model)

NielsRogge commented on docs/source/en/model_doc/cvt.mdx:
```diff
@@ -0,0 +1,57 @@
+
+The CvT model was proposed in [CvT: Introducing Convolutions to Vision Transformers](https://arxiv.org/abs/2103.15808) by Haiping Wu, Bin Xiao, Noel Codella, Mengchen Liu, Xiyang Dai, Lu Yuan and Lei Zhang. The Convolutional vision Transformer (CvT) improves the [Vision Transformer (ViT)](vit) in performance and efficiency by introducing convolutions into ViT to yield the best of both designs.
+
+
+
```
```suggestion
```

NielsRogge commented on docs/source/en/model_doc/cvt.mdx:
```diff
@@ -0,0 +1,57 @@
+Tips:
+
+- CvT models are regular Vision Transformers, but trained with convolutions. They outperform the [original model (ViT)](vit) when fine-tuned on ImageNet-1K and CIFAR-100. You can check out demo notebooks regarding inference as well as fine-tuning on custom data [here](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/VisionTransformer) (you can just replace [`ViTFeatureExtractor`]  ...
+- The available checkpoints are either (1) pre-trained on [ImageNet-22k](http://www.image-net.org/) (a collection of
+
+  14 million images and 22k classes) only, (2) also fine-tuned on ImageNet-22k or (3) also fine-tuned on [ImageNet-1k](http://www.image-net.org/challenges/LSVRC/2012/) (also referred to as ILSVRC 2012, a collection of 1.3 million
+  images and 1,000 classes).
```
```suggestion
- The available checkpoints are either (1) pre-trained on [ImageNet-22k](http://www.image-net.org/) (a collection of 14 million images and 22k classes) only, (2) also fine-tuned on ImageNet-22k or (3) also fine-tuned on [ImageNet-1k](http://www.image-net.org/challenges/LSVRC/2012/) (also referred to as ILSVRC 2012, a collection of 1.3 million images and 1,000 classes).
```

NielsRogge commented on docs/source/en/serialization.mdx:
```diff
@@ -44,7 +44,6 @@ and are designed to be easily extendable to other architectures.
 Ready-made configurations include the following architectures:
 
 <!--This table is automatically generated by `make fix-copies`, do not fill manually!-->
-
```
Normally, this file (serialization.mdx) shouldn't be changed.

NielsRogge commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -309,6 +310,7 @@
         ),
         ("swin", "SwinForImageClassification"),
         ("convnext", "ConvNextForImageClassification"),
+        ("cvt", "CvtForImageClassification"),
```
You can also add CvT to the feature_extraction_auto.py file, such that people will be able to do:

```
from transformers import AutoFeatureExtractor

feature_extractor = AutoFeatureExtractor.from_pretrained("microsoft/cvt-...")
```
You can map "cvt" to `ViTFeatureExtractor`.

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/__init__.py:
```diff
@@ -0,0 +1,61 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2021 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+# coding=utf-8
+# Copyright 2022 AnugunjNaman and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Sorry, we usually don't do this, your name is included in the documentation page ("this model was contributed by...").

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Cvt model configuration"""
```
```suggestion
""" CvT model configuration"""
```

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
```
```suggestion
```

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+
+class CvtConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`CvtModel`]. It is used to instantiate a Cvt model
```
```suggestion
    This is the configuration class to store the configuration of a [`CvtModel`]. It is used to instantiate a CvT model
```

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+    r"""
+    This is the configuration class to store the configuration of a [`CvtModel`]. It is used to instantiate a Cvt model
+    according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the Cvt
```
```suggestion
    defaults will yield a similar configuration to that of the CvT
```

NielsRogge commented on src/transformers/models/cvt/convert_cvt_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,352 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Convert Cvt checkpoints from the original repository.
```
```suggestion
"""Convert CvT checkpoints from the original repository.
```

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+
+logger = logging.get_logger(__name__)
+
+# General docstring
+_CONFIG_FOR_DOC = "CvtConfig"
+_FEAT_EXTRACTOR_FOR_DOC = "AutoFeatureExtractor"
+
+# Base docstring
+_CHECKPOINT_FOR_DOC = "anugunj/cvt-13"
+_EXPECTED_OUTPUT_SHAPE = [1, 384, 14, 14]
+
+# Image classification docstring
+_IMAGE_CLASS_CHECKPOINT = "anugunj/cvt-13"
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tabby, tabby cat"
```
You can also add CvT to the doc tests, as explained here: 

https://github.com/huggingface/transformers/tree/main/docs#testing-documentation-examples

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+
+
+CVT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "anugunj/cvt-13",
+    "anugunj/cvt-13-384-1k",
+    "anugunj/cvt-13-384-22k",
+    "anugunj/cvt-21",
+    "anugunj/cvt-21-384-1k",
+    "anugunj/cvt-21-384-22k",
+    # See all Cvt models at https://huggingface.co/models?filter=cvt
```
To be updated to "microsoft"

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+    return (x, x)
+
+
+# Stochastic depth implementation
+# Taken from https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/drop.py
```
You can add a #copied from statement here, (to be copied from the code in `modeling_convnext.py`)

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+        return hidden_state
+
+
+class ConvEmbeddings(nn.Module):
```
```suggestion
class CvtConvEmbeddings(nn.Module):
```

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+            dropout_rate=config.drop_rate[self.stage],
+        )
+
+        dpr = [x.item() for x in torch.linspace(0, config.drop_path_rate[self.stage], config.depth[stage])]
```
We can rename this to `drop_path_rates` to be clearer

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+            layer_outputs = layer(hidden_state, height, width)
+            hidden_state = layer_outputs
+
+        if self.config.cls_token[self.stage]:
+            cls_token, hidden_state = torch.split(hidden_state, [1, height * width], 1)
```
I would add the CLS token by default to the model outputs, this makes sure no custom output class needs to be created.

NielsRogge reviewed (commented):
Thanks for your PR, it's already in a very good state, some last comments.

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+
+
+    Args:
+        image_size (`int`, *optional*, defaults to 224):
+            The size (resolution) of each image.
```
It looks like this argument isn't used by the model, you can remove it

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+            The size (resolution) of each image.
+        num_channels (`int`, *optional*, defaults to 3):
+            The number of input channels.
+        num_stages (`int`, *optional*, defaults to 3):
+            The number of encoder blocks (i.e. stages in the Mix Transformer encoder).
```
The number of stages can be derived from `len(config.depth)` for instance, not sure if this attribute is required

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+            The dropout probability for stochastic depth, used in the blocks of the Transformer encoder.
+        qkv_bias (`List[bool]`, *optional*, defaults to [True, True, True]]):
+            The bias bool for query, key and value in attentions
+        cls_token (`List[bool]`, *optional*, defaults to [False, False, True]]):
+            The bool for classification token
+        pos_embed (`List[bool]`, *optional*, defaults to [False, False, True]]):
+            The bool for position embeddings
+        cls_token (`List[bool]`, *optional*, defaults to [False, False, True]]):
+            The bool for cls_token
```
`cls_token` is duplicated here.

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
````diff
@@ -0,0 +1,159 @@
+            The standard deviation of the truncated_normal_initializer for initializing all weight matrices.
+        layer_norm_eps (`float`, *optional*, defaults to 1e-6):
+            The epsilon used by the layer normalization layers.
+        Example:
+
+    ```python
+    >>> from transformers import CvtModel, CvtConfig
+
+    >>> # Initializing a Cvt msft/cvt style configuration
+    >>> configuration = CvtConfig()
+
+    >>> # Initializing a model from the msft/cvt style configuration
+    >>> model = CvtModel(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
````
The "Example:" should be shifted to the left (see another configuration as an example), also with an added enter

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/configuration_cvt.py:
```diff
@@ -0,0 +1,159 @@
+
+    def __init__(
+        self,
+        image_size=224,
```
To be removed

AnugunjNaman replied:
Needed for image size during conversion of weight scripts

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+    return output
+
+
+class CvtDropPath(nn.Module):
```
You can also add a copied from statement here:

`# Copied from transformers.models.convnext.modeling_convnext.ConvNextDropPath`

AnugunjNaman replied:
Done

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+    def __init__(
+        self,
+        config,
+        stage,
+    ):
+        super().__init__()
+        self.config = config
+        self.stage = stage
+        if self.config.cls_token[self.stage]:
+            self.cls_token = nn.Parameter(torch.zeros(1, 1, self.config.embed_dim[-1]))
```
Instead of working with a "stage" argument here, maybe we can just do it like this:
```suggestion
        with_cls_token: bool = False,
    ):
        super().__init__()
        self.config = config
        self.stage = stage
        if with_cls_token:
            self.cls_token = nn.Parameter(torch.zeros(1, 1, self.config.embed_dim[-1]))
```

AnugunjNaman replied:
can't eliminate self.stage variable, since it used to get index of all other variables in config like patch_size, padding etc

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+        super().__init__()
+        self.config = config
+        self.stages = nn.ModuleList([])
+        for stage_idx in range(len(config.depth)):
+            self.stages.append(CvtStage(config, stage_idx))
```
```suggestion
        for stage_idx in range(len(config.depth)):
            self.stages.append(CvtStage(config, config.cls_token[stage_idx])
```

NielsRogge commented on src/transformers/models/cvt/modeling_cvt.py:
```diff
@@ -0,0 +1,760 @@
+    hidden_states: Optional[Tuple[torch.FloatTensor]] = None
+
+
+# Copied from transformers.models.vit.modeling_vit.to_2tuple
+def to_2tuple(x):
+    if isinstance(x, collections.abc.Iterable):
+        return x
+    return (x, x)
```
You can remove this `to_2tuple` function, and directly write `x if isinstance(x, collections.abc.Iterable) else (x, x)`

AnugunjNaman replied:
Done

AnugunjNaman committed:
Update README.md

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update docs/source/en/model_doc/cvt.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update docs/source/en/model_doc/cvt.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update docs/source/en/model_doc/cvt.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/__init__.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/configuration_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/configuration_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/configuration_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/configuration_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/configuration_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/convert_cvt_original_pytorch_checkpoint_to_pytorch.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
Update src/transformers/models/cvt/modeling_cvt.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

AnugunjNaman committed:
PR Changes

AnugunjNaman committed:
PR Changes

AnugunjNaman committed:
style

AnugunjNaman committed:
style

AnugunjNaman committed:
cvtmodel forward call doc correction

AnugunjNaman committed:
changes
